pkg/platform/runtime/envdef/file_transform.go

```diff
@@ -0,0 +1,144 @@
+package envdef
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/fileutils"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+)
+
+// FileTransform specifies a single transformation to be performed on files in artifacts post-installation
+type FileTransform struct {
+	Pattern         string           `json:"pattern"`
+	In              []string         `json:"in"`
+	With            string           `json:"with"`
+	ConstTransforms []ConstTransform `json:"const_transforms"`
+	PadWith         *string          `json:"pad_with"`
+}
+
+// ConstTransform is a transformation that should be applied to substituted constants prior to substitution in files
+type ConstTransform struct {
+	In      []string `json:"in"` // List of constants to apply this transform to
+	Pattern string   `json:"pattern"`
+	With    string   `json:"with"`
+}
+
+// applyConstTransforms applies the constant transforms to the Constants values
+func (ft *FileTransform) applyConstTransforms(constants Constants) (Constants, error) {
+	cs := constants
+	for _, ct := range ft.ConstTransforms {
+		fmt.Printf("pattern=%q\n", ct.Pattern)
+		r, err := regexp.Compile(regexp.QuoteMeta(ct.Pattern))
+		if err != nil {
+			return cs, errs.Wrap(err, "Failed to compile regexp pattern in const_transform.")
+		}
+		for _, inVar := range ct.In {
+			inSubst, ok := cs[inVar]
+			if !ok {
+				return cs, errs.New("Do not know what to replace constant %s with.", inVar)
+			}
+			cs[inVar] = r.ReplaceAllString(inSubst, ct.With)
+		}
+	}
+
+	return cs, nil
+}
+
+func (ft *FileTransform) relocateFile(fileBytes []byte, replacement string) ([]byte, error) {
+	findBytes := []byte(ft.Pattern)
+	replacementBytes := []byte(replacement)
+
+	if ft.PadWith == nil {
+		return bytes.ReplaceAll(fileBytes, findBytes, replacementBytes), nil
+	}
+
+	// padding should be one byte
+	if len(*ft.PadWith) != 1 {
+		return fileBytes, errs.New("Padding character needs to have exactly one byte, got %d", len(*ft.PadWith))
```

Comment:
Is this what Shaun said? If not, can you get him to verify this.

Reply:
I put a question in the artifact document.  If the padding character is wider, it may be impossible to ensure that the relocated string has the same length as the original.